pkg/operator/configobservation/encryption/shared_encryption.go

```diff
@@ -0,0 +1,395 @@
+
+func appendSecretPerAnnotationState(in, out *[]*corev1.Secret, secret *corev1.Secret, annotation string) {
+	if len(secret.Annotations[annotation]) != 0 {
+		*in = append(*in, secret)
+	} else {
+		*out = append(*out, secret)
+	}
+}
+
+func secretToKey(encryptionSecret *corev1.Secret, validGRs map[schema.GroupResource]bool) (schema.GroupResource, apiserverconfigv1.Key, uint64, bool) {
+	group := encryptionSecret.Labels[encryptionSecretGroup]
+	resource := encryptionSecret.Labels[encryptionSecretResource]
+	keyData := encryptionSecret.Data[encryptionSecretKeyData]
+
+	keyID, validKeyID := secretToKeyID(encryptionSecret)
+
+	gr := schema.GroupResource{Group: group, Resource: resource}
+	key := apiserverconfigv1.Key{
+		// limit the length of the name as it is used as a prefix for every value in etcd
+		// this means that each resource can have 100 active keys (0 - 99 so two ASCII letters max)
+		// thus to avoid collisions something must prune the old ones (that is fine since we need pruning anyway)
+		Name:   strconv.FormatUint(keyID%100, 10),
+		Secret: base64.StdEncoding.EncodeToString(keyData),
+	}
+	invalidKey := len(resource) == 0 || len(keyData) == 0 || !validKeyID || !validGRs[gr]
+
+	return gr, key, keyID, !invalidKey
+}
+
+func secretToKeyID(encryptionSecret *corev1.Secret) (uint64, bool) {
+	// see format and ordering comment at top
+	lastIdx := strings.LastIndex(encryptionSecret.Name, "-")
+	keyIDStr := encryptionSecret.Name[lastIdx+1:]
+	keyID, keyIDErr := strconv.ParseUint(keyIDStr, 10, 0)
+	invalidKeyID := lastIdx == -1 || keyIDErr != nil
+	return keyID, !invalidKeyID
+}
+
+func getResourceConfigs(encryptionState groupResourcesState) []apiserverconfigv1.ResourceConfiguration {
+	resourceConfigs := make([]apiserverconfigv1.ResourceConfiguration, 0, len(encryptionState))
+
+	for gr, grKeys := range encryptionState {
+		resourceConfigs = append(resourceConfigs, apiserverconfigv1.ResourceConfiguration{
+			Resources: []string{gr.String()}, // we are forced to lose data here because this API is broken
+			Providers: keysToProviders(grKeys),
+		})
+	}
+
+	// make sure our output is stable
+	sort.Slice(resourceConfigs, func(i, j int) bool {
+		return resourceConfigs[i].Resources[0] < resourceConfigs[j].Resources[0] // each resource has its own keys
+	})
+
+	return resourceConfigs
+}
+
+func grKeysToDesiredKeys(grKeys keysState) desiredKeys {
+	desired := desiredKeys{}
+
+	desired.writeKey, desired.hasWriteKey = determineWriteKey(grKeys)
+
+	// keys have a duplicate of the write key
+	// or there is no write key
+
+	// iterate in reverse to order the read keys in optimal order
+	for i := len(grKeys.keys) - 1; i >= 0; i-- {
+		readKey := grKeys.keys[i]
+		if desired.hasWriteKey && readKey == desired.writeKey {
+			continue // if present, drop the duplicate write key from the list
+		}
+		desired.readKeys = append(desired.readKeys, readKey)
+	}
+
+	return desired
+}
+
+func determineWriteKey(grKeys keysState) (apiserverconfigv1.Key, bool) {
```

Comment:
If I am not mistaken here, this func mutate `grKeys`. That's bad. Maybe make it explicit as a method of `grKeys`.

Reply:
Uh where does it mutate?